blog/content/post/2020-07-06-qubesos-fwupd-core.md

```diff
@@ -0,0 +1,191 @@
+the JSON form. 1.2.12 version has no such command. At first, we need to acquire
+devices information from `fwupdagent get-devices`. Updates details are obtained
+in the comparison of the current device version with versions in a release
+list.
+
+### qubes-fwupdmgr update
+
+During the update process, a user gets a list of devices with information about
+possible updates. The user chooses the device number, which will be updated to
+the newest version. Then the tool downloads, validates, and installs firmware
+update `.cab` archive.
+
+![qubes_fwupdmgr-update](/img/qfwupd-update.png)
+
+Downloading a firmware archive and the metadata updates vary in the verification
+process. `fwupd-download-updates.sh` checks the checksum of the archive in the
+updateVM. Then the `fwupd_receive_updates` copies the firmware to dom0 and it
+verifies the checksum once again. The script unpacks the archive and checks the
+signature of the firmware. Also during the system firmware update,
+`qubes_fwupdmgr` verifies DMI details with the information contained in firmware
+metadata. If the device is connected in usbVM, `qubes_fwupdmgr` copies
+the firmware file to sys-usb. The main script initializes
+`fwupd_usbvm_validate`, that verifies the firmware in usbVM before it will be
+installed.
+
+The update process of the device connected in dom0 is shown below.
+
+[![asciicast](https://asciinema.org/a/cv2Iyv10EkF9lqrRtfkVLvGja.svg)](https://asciinema.org/a/cv2Iyv10EkF9lqrRtfkVLvGja)
+
+Here is the update process of the device connected in sys-usb.
+
+[![asciicast](https://asciinema.org/a/v7ZiSG3Xp9fauzwFkMNZgmHgt.svg)](https://asciinema.org/a/v7ZiSG3Xp9fauzwFkMNZgmHgt)
+
+### qubes-fwupdmgr downgrade
+
+The downgrade process is similar to updating. A user chooses the device number
+and the firmware version, that will be installed on the downgraded device. Then
+the tool downloads, validates, and installs firmware downgrade `.cab` archive.
+The firmware installation proceeds with `--allow-older` flag.
+
+The downgrade process is presented in the following video.
+
+![downgrade-usbvm](/img/qfwupd-downgrade.png)
+
+### qubes-fwupdmgr clean
+
+The `clean` command removes cache directories from dom0 and updateVM.
+
+[![asciicast](https://asciinema.org/a/0ZT3Gi2SzcPxUMWgNA56BPYzC.svg)](https://asciinema.org/a/0ZT3Gi2SzcPxUMWgNA56BPYzC)
+
+## Tests
+
+The `qubes_fwupdmgr.py` is covered by python tests. They could be run on any OS.
+The unit tests and integration tests, that require Qubes OS or testing device
+will be skipped if the requirements aren't met.
+
+Following videos show the testing process on Qubes OS and Ubuntu.
+
+[![asciicast](https://asciinema.org/a/TgHOkLnD2YICxB0U80PVcQGqX.svg)](https://asciinema.org/a/TgHOkLnD2YICxB0U80PVcQGqX)
+
+[![asciicast](https://asciinema.org/a/pafnoJp50uQj0qKESCghS4FYW.svg)](https://asciinema.org/a/pafnoJp50uQj0qKESCghS4FYW)
+
+## Development roadmap
+
+The next step of the development process will be replacing the static paths of
```

Comment:
@Asiderr I would add a little bit more description of this issue with paths to binaries being changes across fwupd releases. Current sentence may not reflect what is the real issue and how we plan to solve it.

Reply:
@macpijan I've added more information on this topic and changed the rest of this paragraph.